ports/alif/machine_pin.c

```diff
@@ -225,6 +288,125 @@ static mp_obj_t machine_pin_toggle(mp_obj_t self_in) {
 }
 static MP_DEFINE_CONST_FUN_OBJ_1(machine_pin_toggle_obj, machine_pin_toggle);
 
+static machine_pin_irq_obj_t *machine_pin_get_irq(machine_pin_obj_t *self) {
+    size_t irq_idx = self->port * MACHINE_PIN_NUM_PORT_IO + self->pin;
+    machine_pin_irq_obj_t *irq = MP_STATE_PORT(machine_pin_irq_obj[irq_idx]);
+
+    // Allocate a new IRQ object if it doesn't exist.
+    if (irq == NULL) {
+        irq = m_new_obj(machine_pin_irq_obj_t);
+        irq->base.base.type = &mp_irq_type;
+        irq->base.methods = (mp_irq_methods_t *)&machine_pin_irq_methods;
+        irq->base.parent = MP_OBJ_FROM_PTR(self);
+        irq->base.handler = mp_const_none;
+        irq->base.ishard = false;
+        irq->irq_num = (self->port < 15) ? (GPIO0_IRQ0_IRQn + irq_idx) : (LPGPIO_IRQ0_IRQn + self->pin);
+        irq->reserved = false;
+        MP_STATE_PORT(machine_pin_irq_obj[irq_idx]) = irq;
+    }
+
+    return irq;
+}
+
+static mp_uint_t machine_pin_irq_trigger(mp_obj_t self_in, mp_uint_t trigger) {
+    machine_pin_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    machine_pin_irq_obj_t *irq = machine_pin_get_irq(self);
```

Comment:
If the IRQ object doesn't exist, then this function should just return 0.  This function should never try to allocate a new object (it might be running in a hard interrupt).

So probably you want a simple `machine_pin_get_irq()` which may return NULL, and `machine_pin_get_irq_or_create()` which creates if it doesn't exist.

Reply:
> This function should never try to allocate a new object (it might be running in a hard interrupt).

I don't think it's possible for this allocate memory because you'd have to have an irq object first to call any function that calls this function, and `machine_pin_get_irq` already allocates the memory. Or am I missing something? It's possible to just add a `bool alloc` but I don't think it's needed.

Reply:
> I don't think it's possible for this function to allocate memory when called from callback, because you'd have to have an irq object first to call any function that calls this function, and `machine_pin_get_irq` already allocates the memory.

That might be true, but IMO it's hard to reason about, especially if you come to this code with fresh eyes.

I suggest making it explicit that these functions should never allocate, eg with a `bool alloc` argument would be fine.

Reply:
Alternatively, I could add a macro to calculate the IRQ index and move the alloc to make_new? The reason for adding this function to begin with is to not have that code repeated everywhere just in case it changes in the future (for another MCU).

Reply:
> I could add a macro to calculate the IRQ index and move the alloc to make_new?

Yes, that would be good.